W-10890554-db-1.13-release-kt (#1859)
mulesoft/docs-connectors: db/1.13/modules/ROOT/pages/database-connector-udt-stored-procedure.adoc

Previous Mule versions provided the `createStruct` and `createArray` functions, which are now deprecated. These functions created JDBC values right away, when they must have been created with the same connection used in the Database Connector operation. The new `prepareStruct` and `prepareArray` functions enable Database Connector to defer these JDBC values creation until the right SQL connection is available.

[%header,%cols]
|===
| Deprecated Function | Active Function
| `Db::createArray(configName, typeName, values): java.sql.Array` | `Db::prepareArray(typeName, values): java.util.List`
| `Db::createStruct(configName, typeName, properties): java.sql.Struct` | `Db::prepareStruct(typeName, properties): java.util.List`
|===

=== Parameters

`Db::prepareArray(typeName, values): java.util.List` function prepares an array for sending to the database as an java.sql.Array.

[%header,%cols]
|===
| Parameter | Type | Description
| `typeName` | String | Name of the Array type to create.
| `values` | List<Object> | Values to pass to the database.
| `java.util.List` | List<Object> | The return value is an immutable list that holds all the information for creating an array later.
|===

> `Db::prepareStruct(typeName, properties): java.util.List` function prepares an struct for sending to the database as a java.sql.Struct.

Reviewer: ```suggestion
`Db::prepareStruct(typeName, properties): java.util.List` function prepares a struct for sending to the database as a java.sql.Struct.
```

[%header,%cols]
|===
| Parameter | Type | Description
| `typeName` | String | Name of the Struct type to create.
| `properties` | List<Object> | Properties to pass to the database.
| `java.util.List` | List<Object> | The return value is an immutable list that holds all the information for creating a struct later.
|===

[NOTE]

> If you have existing Mule apps that use the deprecated functions, replace  `createStruct` with `prepareStruct` and `createArray` with `prepareArray`. Additionally, remove the first parameter `configName` from the function structure.

Reviewer: ```suggestion
If you have existing Mule apps that use the deprecated functions, replace `createStruct` with `prepareStruct` and `createArray` with `prepareArray`. Additionally, remove the first parameter `configName` from the function structure.
```

Oracle enables you to create your custom data types, known as _user-defined types (UDTs)_. The following example shows how to configure Anypoint Connector for Database (Database Connector) *Stored procedure* operation in a Mule application to call a stored procedure that uses UDTs from an Oracle database. +
The stored procedure performs two actions: +

* Inserts the first name, last name, age, and phone number of people whose age is less than 100 into a table
* Returns a list of phone numbers of all the people added to the table

To accomplish this example, you must: +

* Run a script to create UDTs, the table, and the stored procedure
* Configure an Oracle database connection
* Create the Mule application
* Use `prepareStruct` and `prepareArray` DataWeave functions to map application data to custom types
* Run and test the application with curl commands

.Oracle Stored procedure with UDTs Flow
image::database-oracle-udt-flow.png[The Oracle UDTs flow in Studio]

== Using prepareStruct and prepareArray Functions

The `createStruct` and `createArray` functions are deprecated and replaced by the `prepareStruct` and `prepareArray` functions.

> Previous Mule versions provided only the `createStruct` and `createArray` functions that created JDBC values right away. However, JDBC values must be created with the same connection used in the Database Connector operation. The `prepareStruct` and `prepareArray` functions enables to defer the JDBC values creation until the right SQL connection is available.

Reviewer: ```suggestion
Previous Mule versions provided only the `createStruct` and `createArray` functions that created JDBC values right away. However, JDBC values must be created with the same connection used in the Database Connector operation. The `prepareStruct` and `prepareArray` functions enable you to defer the JDBC values creation until the right SQL connection is available.
```

[%header,%cols]
|===
| Deprecated Function | Active Function
| `Db::createArray(configName, typeName, values): java.sql.Array` | `Db::prepareArray(typeName, values): java.util.List`
| `Db::createStruct(configName, typeName, properties): java.sql.Struct` | `Db::prepareStruct(typeName, properties): java.util.List`
|===

=== Parameters

`Db::prepareArray(typeName, values): java.util.List` function prepares an array for sending to the database as an java.sql.Array.

[%header,%cols]
|===
| Parameter | Type | Description
| `typeName` | String | Name of the Array type to create.
| `values` | List<Object> | Values to pass to the database.
| `java.util.List` | List<Object> | The return value is an immutable list that holds all the information for creating an array later.
|===

> `Db::prepareStruct(typeName, properties): java.util.List` function prepares an struct for sending to the database as an java.sql.Struct.

Reviewer: ```suggestion
`Db::prepareStruct(typeName, properties): java.util.List` function prepares a struct for sending to the database as an java.sql.Struct.
```

[%header,%cols]
|===
| Parameter | Type | Description
| `typeName` | String | Name of the Struct type to create.
| `properties` | List<Object> | Properties to pass to the database.

> | `java.util.List` | List<Object> | The return value is an immutable list that holds all the information for creating an struct later.

Reviewer: ```suggestion
| `java.util.List` | List<Object> | The return value is an immutable list that holds all the information for creating a struct later.
```

.Transform Message configuration
image::database-oracleudt-example-6.png[Transform Message configuration]

> When the Mule app performs a Database Connector operation, the `prepareStruct` instructs Mule to create JDBC struct objects based on the `Name` type and their corresponding properties. In the example, types `PERSON` and `PHONE_NUMBER` are struct objects. The function expects the following parameters in order:

Reviewer 2: This doesn't happen when the Mule app performs an operation, is a DW function used to "prepare" data that will be sent to the db at a later time. The original text tried to account for this difference.
